src/cluster.c

```diff
@@ -5110,6 +5248,12 @@ void clusterReplyShards(client *c) {
      * information and an empty slots array. */
     while((de = dictNext(di)) != NULL) {
         clusterNode *n = dictGetVal(de);
+        /* Failed primary nodes will lose their slots after automatic
+         * failover. They will be picked up by the new primary later
+         * so it is OK to skip them */
+        if (nodeFailed(n) && n->slot_info_pairs_count == 0) {
+            continue;
+        }
```

Comment:
If the node just failed and no new primary has been elected, its slot count should be non-zero (assuming non-empty shard) so it will get picked up.

I like your suggestion of refactoring this loop to go over the shards dictionary instead so this non-so-intuitive check can be removed I believe